### PR TITLE
Fix spectral illuminance to CIE XYZ conversions

### PR DESCRIPTION
**Pull request by mororo250**
Hey guys, this is what I was working at before I stopped last year ...
This basically fix the green/blue tints when physically rendering sky:

appleseed

This also improve conversion from cieXYZ to spectrum.

I changed the cornell box light radiance to the one used in: https://www.graphics.cornell.edu/online/box/data.html, I don't know where does your current data come from. But it got I bit reddish after the change:

1_appleseed_new

Using data from https://www.graphics.cornell.edu/online/box/data.html,:
1_appleseed_new_original_values

### CODE REVIEW

File: src/appleseed/renderer/modeling/input/colorsource.cpp [hunk 1]
```diff
@@ -53,14 +53,14 @@ namespace renderer
...
-ColorSource::ColorSource(const ColorEntity& color_entity)
+ColorSource::ColorSource(const ColorEntity& color_entity, const char* intent)
   : Source(true)
   , m_color_entity(color_entity)
 {
     // Retrieve the color values.
     if (color_entity.get_color_space() == ColorSpaceSpectral)
-        initialize_from_spectrum(color_entity);
-    else initialize_from_color3(color_entity);
+        initialize_from_spectrum(color_entity, std::strcmp(intent, "reflectance") == 0 ? Spectrum::Reflectance : Spectrum::Illuminance);
```
Comment (mororo250) [inline src/appleseed/renderer/modeling/input/colorsource.cpp]: I don't like it. I will try to do some refactoring and remove this string comparison.

File: src/appleseed/renderer/modeling/surfaceshader/constantsurfaceshader.cpp [hunk 2]
```diff
@@ -120,7 +120,7 @@ namespace
                 &values);
 
             // Initialize the shading result.
-            shading_result.m_main.rgb() = values.m_color.to_rgb(g_std_lighting_conditions);
+            shading_result.m_main.rgb() = values.m_color.reflectance_to_rgb(g_std_lighting_conditions);
```
Comment (mororo250) [inline src/appleseed/renderer/modeling/surfaceshader/constantsurfaceshader.cpp]: In some cases like this one. I couldn't tell for sure if it was an emission spectrum of a reflective spectrum. I was hoping to get some help, so I wouldn't need to dig the source code too much.

**Comment by mororo250** [comment]
Don't review this yet.

**Comment by dictoon** [comment]
> I changed the cornell box light radiance to the one used in: https://www.graphics.cornell.edu/online/box/data.html, I don't know where does your current data come from. But it got I bit reddish after the change:

We used the official light radiance from that website, just resampled to the [400, 700] nm range with 10 nm steps.

**Comment by dictoon** [comment]
> > I changed the cornell box light radiance to the one used in: https://www.graphics.cornell.edu/online/box/data.html, I don't know where does your current data come from. But it got I bit reddish after the change:
>
> We used the official light radiance from that website, just resampled to the [400, 700] nm range with 10 nm steps.

Actually:

Our canonical Cornell Box dataset is in `renderer/modeling/project-builtin/cornellboxproject.cpp`. The test scene you modified is just one of many Cornell Box scenes in the test suite.

And you're right, we're not using the 4-value, [400, 700] nm spectrum that you used for the light source radiance. I believe we're using this spectrum instead:
https://www.graphics.cornell.edu/online/measurements/source-spectra/index.html

Our renders match the reference renders and photography fairly well. Not sure why you got a much brighter result, you shouldn't.

**Comment by dictoon** [comment]
Maybe a normalization issue?

**Comment by mororo250** [comment]
This is now ready for review.

**Comment by mororo250** [comment]
I changed the light radiance on the Cornell Box to the CIE Standard Illuminant D65 and got the follows XYZ tristimulus values :
After the change: 93.9054718 98.9102402 107.524445
Before: 90.2572708 99.6380539 118.463623
Expected: 95.047 100.00 108.883 

The results after the change are far closer to the expected one. The difference is probably due to the spectral range we use.

**Comment by mororo250** [comment]
before:
d65_old_rgb
After:
d65_rgb

**Comment by mororo250** [comment]
There was a normalization issue indeed:

File: src/appleseed/renderer/modeling/surfaceshader/constantsurfaceshader.cpp (hunk 2 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/surfaceshader/constantsurfaceshader.cpp:123]: I think we should consider this an emission since loosely speaking, it produces "colors" that are directly visible "on screen".

**Comment by dictoon** [comment]
Thanks for the really nice work @mororo250!

Obviously this part of the code is rather sensitive, and even though what we have right now in master is almost certainly either not entirely correct or optimal in any sense, we need to be careful when changing it in order to avoid introducing new regressions.

Could you show again:
- A comparison of the built-in Cornell Box scene, before and after your changes?
- A comparison of the house with blue sky scene, before and after your changes?

**Comment by dictoon** [comment]
I would also appreciate if you could run the full test suite on this PR and check the results. The code you propose to change is at the core of the renderer and has the capacity to introduce severe regressions.

**Comment by mororo250** [comment]
Sorry, I didn't see your comment here. I will do it.

**Comment by mororo250** [comment]
appleseed_house_old

appleseed_house_new

**Comment by mororo250** [comment]
appleseed_old

appleseed_new

**Comment by dictoon** [comment]
Thanks! I assume the first pic is before and the second one is after?

**Comment by mororo250** [comment]
The first pic is before and the second is after.

**Comment by mororo250** [comment]
Link for download the test suite: https://wetransfer.com/downloads/6c2fffa1fd9a85181539cdde11b1a11420200418164403/665c44e90293961187fab2f85c6026cc20200418164619/6f8250?utm_campaign=WT_email_tracking&utm_content=general&utm_medium=download_button&utm_source=notify_recipient_email

File: src/appleseed/foundation/image/colorspace.h [hunk 3]
```diff
@@ -863,10 +884,36 @@ Color<T, 3> spectrum_to_ciexyz(
...
+        "foundation::spectrum_illuminance_to_ciexyz() expects 31-channel spectra");
+
+    T x = T(0.0);
+    T y = T(0.0);
+    T z = T(0.0);
+
+    for (size_t w = 0; w < 31; ++w)
+    {
+        const T val = spectrum[w];
+        x += XYZCMFCIE19312Deg[0][w] * val;
```
[the remark this answers is not in the document]
Comment (mororo250) [reply src/appleseed/foundation/image/colorspace.h:901]: A reflectance spectrum is not directly converted to another color space. We convert a reflected spectrum that we get when the reflectance spectrum is illuminated by a certain illuminant.
LightingConditions specify what that illuminant would be. I don't think it makes much sense to use it here since an illuminance spectrum doesn't need any lighting condition.

[the remark this answers is not in the document]
Comment (mororo250) [reply src/appleseed/foundation/image/colorspace.h:901]: Maybe I should abstract XYZCMFCIE19312Deg into another class?

File: src/appleseed/foundation/image/colorspace.h [hunk 4]
```diff
@@ -335,10 +352,14 @@ T luminance(const Color<T, 3>& linear_rgb);
 
 // Convert a spectrum to a color in the CIE XYZ color space.
 template <typename T, typename SpectrumType>
-Color<T, 3> spectrum_to_ciexyz(
+Color<T, 3> spectrum_reflectance_to_ciexyz(
     const LightingConditions&   lighting,
     const SpectrumType&         spectrum);
 
+template <typename T, typename SpectrumType>
+Color<T, 3> spectrum_illuminance_to_ciexyz(const SpectrumType& spectrum);
```
[the remark this answers is not in the document]
Comment (mororo250) [reply src/appleseed/foundation/image/colorspace.h:360]: spectral_illuminance_to_ciexyz()?

File: src/appleseed/foundation/image/colorspace.h [hunk 5]
```diff
@@ -863,10 +884,36 @@ Color<T, 3> spectrum_to_ciexyz(
...
+
+    for (size_t w = 0; w < 31; ++w)
+    {
+        const T val = spectrum[w];
+        x += XYZCMFCIE19312Deg[0][w] * val;
+        y += XYZCMFCIE19312Deg[1][w] * val;
+        z += XYZCMFCIE19312Deg[2][w] * val;
+    }
+
+    x *= NORMALIZATIONCIE19312Deg;
```
[the remark this answers is not in the document]
Comment (mororo250) [reply src/appleseed/foundation/image/colorspace.h:906]: We should rename LightningCoditions then. Since spectral illuminance doesn't need any lighting condition.

File: src/appleseed/renderer/modeling/input/inputarray.h [hunk 6]
```diff
@@ -53,7 +53,7 @@ namespace renderer
 // Input formats.
 //
 
-enum InputFormat
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/input/inputarray.h:56]: That's a good change but it obscures the main changes of this PR. If possible, I would suggest to split this into its own commit.

File: src/appleseed/renderer/modeling/input/inputarray.h [hunk 7]
```diff
@@ -53,7 +53,7 @@ namespace renderer
...
+enum class InputFormat
 {
     InputFormatFloat,
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/input/inputarray.h:58]: If the enum is scoped, the `InputFormat` prefixes become redundant and should be removed.

File: src/appleseed/renderer/modeling/input/inputarray.h (hunk 6 above)
Comment (mororo250) [reply src/appleseed/renderer/modeling/input/inputarray.h:56]: it won't compile

File: src/appleseed/renderer/utility/dynamicspectrum.h [hunk 8]
```diff
@@ -470,27 +519,48 @@ inline const T& DynamicSpectrum<T, N>::operator[](const size_t i) const
...
+inline foundation::Color<T, 3> DynamicSpectrum<T, N>::illuminance_to_ciexyz(const foundation::ColorMatchingFunction& cmf) const
+{
+    return
+        s_mode == RGB
+        ? linear_rgb_to_ciexyz(
+            foundation::Color<T, 3>(m_samples[0], m_samples[1], m_samples[2]))
+        : foundation::spectral_illuminance_to_ciexyz<T>(cmf, *this);
 }
 
+
```
Comment (dictoon) [inline src/appleseed/renderer/utility/dynamicspectrum.h:563]: Remove extra blank line.

File: src/appleseed/renderer/utility/dynamicspectrum.h [hunk 9]
```diff
@@ -126,19 +130,31 @@ class DynamicSpectrum
...
+    foundation::Color<ValueType, 3> reflectance_to_rgb(
         const foundation::LightingConditions&   lighting_conditions) const;
 
+    // Use in emissive samples.
+    foundation::Color<ValueType, 3> illuminance_to_rgb(const foundation::ColorMatchingFunction& cmf) const;
+
     // Convert the spectrum to a CIE XYZ color.
-    foundation::Color<ValueType, 3> to_ciexyz(
+    // Use in reflective and transmissive samples
+    foundation::Color<ValueType, 3>  reflectance_to_ciexyz(
```
Comment (dictoon) [inline src/appleseed/renderer/utility/dynamicspectrum.h:152]: Remove extra space after type.

File: src/appleseed/renderer/utility/dynamicspectrum.h [hunk 10]
```diff
@@ -126,19 +130,31 @@ class DynamicSpectrum
...
+    // Use in reflective and transmissive samples.
+    foundation::Color<ValueType, 3> reflectance_to_rgb(
         const foundation::LightingConditions&   lighting_conditions) const;
 
+    // Use in emissive samples.
+    foundation::Color<ValueType, 3> illuminance_to_rgb(const foundation::ColorMatchingFunction& cmf) const;
+
     // Convert the spectrum to a CIE XYZ color.
-    foundation::Color<ValueType, 3> to_ciexyz(
+    // Use in reflective and transmissive samples
```
Comment (dictoon) [inline src/appleseed/renderer/utility/dynamicspectrum.h:151]: End full line comments with periods.

File: src/appleseed/renderer/modeling/input/colorsource.h [hunk 11]
```diff
@@ -32,6 +32,7 @@
 // appleseed.renderer headers.
 #include "renderer/global/globaltypes.h"
 #include "renderer/modeling/input/source.h"
+#include "renderer/modeling/input/inputarray.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/input/colorsource.h:35]: Order alphabetically.

File: src/appleseed/renderer/modeling/input/colorsource.cpp [hunk 12]
```diff
@@ -36,6 +36,7 @@
 #include "renderer/modeling/color/wavelengths.h"
 #include "renderer/modeling/entity/entity.h"
 
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/input/colorsource.cpp:39]: Remove redundant blank line.

File: src/appleseed/renderer/modeling/color/colorspace.h [hunk 13]
```diff
@@ -34,10 +34,18 @@
...
+//
+// Standard lighting conditions (D65 illuminant, CIE 1931 2-deg color matching functions).
+//
+
+extern foundation::LightingConditions       g_std_lighting_conditions;
+
+
 //
-// Standard lighting conditions (D65 illuminant, CIE 1964 10-deg color matching functions).
+// Standard Color matching function CIE 1931 2-deg
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/color/colorspace.h:46]: Choose your casing: either "Standard Color Matching Functions" or "Standard color matching functions".

File: src/appleseed/renderer/modeling/color/colorspace.h [hunk 14]
```diff
@@ -34,10 +34,18 @@
 namespace renderer
 {
 
+
+//
+// Standard lighting conditions (D65 illuminant, CIE 1931 2-deg color matching functions).
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/color/colorspace.h:39]: The previous comment stated the use of CIE 1964 10-deg CMFs, now your comments states the use of CIE 1931 2-deg CMFs. Did you change the standard lighting conditions?

File: src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp [hunk 15]
```diff
@@ -171,7 +171,7 @@ namespace
     {
         return std::accumulate(retAp.begin(), retAp.end(), 0.0f, [](float s, const Spectrum& ap)
         {
-            return s + luminance(ap.to_rgb(g_std_lighting_conditions));
+            return s + luminance(ap.illuminance_to_ciexyz(g_std_cmf));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp:174]: Regression, you're now computing luminance on a CIE XYZ triplet. Please double-check the rest of the PR for similar regressions (because I assume they may have been introduced with global search & replace operations).

File: src/appleseed/foundation/image/colorspace.h [hunk 16]
```diff
@@ -171,6 +173,23 @@ class LightingConditions
 };
 
 
+//
+// Color matching functions and its normalization factor.
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:177]: its -> their

File: src/appleseed/foundation/image/colorspace.h [hunk 17]
```diff
@@ -171,6 +173,23 @@ class LightingConditions
...
+// Used in spectral illuminance conversion.
+//
+
+class ColorMatchingFunction
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:181]: It should be `ColorMatchingFunctions` (plural). Comments should also use the plural form.

File: src/appleseed/foundation/image/colorspace.h [hunk 18]
```diff
@@ -335,13 +354,22 @@ T luminance(const Color<T, 3>& linear_rgb);
...
 // Convert a spectrum to a color in the CIE XYZ color space.
 template <typename T, typename SpectrumType>
-Color<T, 3> spectrum_to_ciexyz(
+Color<T, 3> spectral_reflectance_to_ciexyz(
     const LightingConditions&   lighting,
     const SpectrumType&         spectrum);
 
+template <typename T, typename SpectrumType>
+Color<T, 3> spectral_illuminance_to_ciexyz(
+    const ColorMatchingFunction& cmf,
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:363]: Align arguments with neighboring functions.

File: src/appleseed/foundation/image/colorspace.h [hunk 19]
```diff
@@ -877,24 +910,41 @@ inline Color3f spectrum_to_ciexyz<float, RegularSpectrum31f>(
...
+Color<T, 3> spectral_reflectance_to_ciexyz(
+    const LightingConditions&   lighting,
+    const SpectrumType&         spectrum)
+{
+    return spectrum_to_ciexyz<T, SpectrumType>(lighting.m_cmf, lighting.m_rcp_n, spectrum);
+}
+
+template <typename T, typename SpectrumType>
+Color<T, 3> spectral_illuminance_to_ciexyz(
+    const ColorMatchingFunction& cmf,
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:942]: IMHO the CMFs should be part of `LightingConditions`. Don't you think?

File: src/appleseed/renderer/modeling/color/colorspace.h (hunk 14 above)
Comment (mororo250) [reply src/appleseed/renderer/modeling/color/colorspace.h:39]: The comment was wrong

File: src/appleseed/foundation/image/colorspace.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (dictoon) [reply src/appleseed/foundation/image/colorspace.h]: Yes absolutely :)

File: src/appleseed/foundation/image/colorspace.h [hunk 20]
```diff
@@ -840,13 +846,13 @@ inline T luminance(const Color<T, 3>& linear_rgb)
 //
 
 template <typename T, typename SpectrumType>
-Color<T, 3> spectrum_to_ciexyz(
-    const LightingConditions&   lighting,
-    const SpectrumType&         spectrum)
+inline Color<T, 3> spectrum_to_ciexyz(
+    const Color4f   cmf[32],
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:850]: Align arguments.

File: src/appleseed/foundation/image/colorspace.h [hunk 21]
```diff
@@ -855,20 +861,19 @@ Color<T, 3> spectrum_to_ciexyz(
...
     return Color<T, 3>(x, y, z);
 }
 
 #ifdef APPLESEED_USE_SSE
-
 template <>
 inline Color3f spectrum_to_ciexyz<float, RegularSpectrum31f>(
-    const LightingConditions&   lighting,
-    const RegularSpectrum31f&   spectrum)
+    const Color4f   cmf[32],
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:875]: Align arguments.

File: src/appleseed/foundation/image/colorspace.h [hunk 22]
```diff
@@ -840,13 +846,13 @@ inline T luminance(const Color<T, 3>& linear_rgb)
...
-    const LightingConditions&   lighting,
-    const SpectrumType&         spectrum)
+inline Color<T, 3> spectrum_to_ciexyz(
+    const Color4f   cmf[32],
+    const SpectrumType& spectrum)
 {
     static_assert(
         SpectrumType::Samples == 31,
-        "foundation::spectrum_to_ciexyz() expects 31-channel spectra");
+        "foundation::spectral_reflectance_to_ciexyz() expects 31-channel spectra");
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:855]: Revert.

File: src/appleseed/foundation/image/colorspace.h [hunk 23]
```diff
@@ -892,9 +897,24 @@ inline Color3f spectrum_to_ciexyz<float, RegularSpectrum31f>(
...
+Color<T, 3> spectral_reflectance_to_ciexyz(
+    const LightingConditions&   lighting,
+    const SpectrumType&         spectrum)
+{
+    return spectrum_to_ciexyz<T, SpectrumType>(lighting.m_cmf_reflectance, spectrum);
+}
+
+template <typename T, typename SpectrumType>
+Color<T, 3> spectral_illuminance_to_ciexyz(
+    const LightingConditions& lighting,
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:912]: Align arguments.

File: src/appleseed/foundation/image/colorspace.h [hunk 24]
```diff
@@ -892,9 +897,24 @@ inline Color3f spectrum_to_ciexyz<float, RegularSpectrum31f>(
 
     return Color3f(transfer[0], transfer[1], transfer[2]);
 }
-
 #endif  // APPLESEED_USE_SSE
 
+template <typename T, typename SpectrumType>
+Color<T, 3> spectral_reflectance_to_ciexyz(
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:903]: Mark as `inline`.

File: src/appleseed/foundation/image/colorspace.h [hunk 25]
```diff
@@ -892,9 +897,24 @@ inline Color3f spectrum_to_ciexyz<float, RegularSpectrum31f>(
...
+    const LightingConditions&   lighting,
+    const SpectrumType&         spectrum)
+{
+    return spectrum_to_ciexyz<T, SpectrumType>(lighting.m_cmf_reflectance, spectrum);
+}
+
+template <typename T, typename SpectrumType>
+Color<T, 3> spectral_illuminance_to_ciexyz(
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:911]: Mark as `inline`.

File: src/appleseed/foundation/image/colorspace.h [hunk 26]
```diff
@@ -161,7 +161,8 @@ extern const RegularSpectrum31f RGBToSpectrumBlueIlluminance;
 class LightingConditions
 {
   public:
-    APPLESEED_SIMD4_ALIGN Color4f   m_cmf[32];                  // precomputed values of (cmf[0], cmf[1], cmf[2]) * illuminant
+    APPLESEED_SIMD4_ALIGN Color4f   m_cmf_reflectance[32];      // precomputed values of (cmf[0], cmf[1], cmf[2]) * illuminant
+    APPLESEED_SIMD4_ALIGN Color4f   m_cmf_illuminance[32];      // Rescale the color matching functions such that luminance integrates to 1.
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.h:165]: I'm not sure the comment here is correct, or at least it's not complete as it doesn't seem to describe what `m_cmf_illuminance` is.

File: src/appleseed/foundation/image/colorspace.cpp [hunk 27]
```diff
@@ -1443,6 +1443,15 @@ const RegularSpectrum31f RGBToSpectrumRedIlluminance(RegularSpectrum31f::from_ar
 const RegularSpectrum31f RGBToSpectrumGreenIlluminance(RegularSpectrum31f::from_array(RGBToSpectrumGreenIlluminanceTab));
 const RegularSpectrum31f RGBToSpectrumBlueIlluminance(RegularSpectrum31f::from_array(RGBToSpectrumBlueIlluminanceTab));
 
+// Get color matching functions normalizer
+float get_cmf_normalizer(Color4f cmf[32])
+{
+    float n = 0.0f;
+    for (std::size_t w = 0; w < 31; ++w)
+        n += cmf[w][1];
+   return 1.0f / n;
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.cpp:1452]: Fix indentation.

File: src/appleseed/foundation/image/colorspace.cpp [hunk 28]
```diff
@@ -1443,6 +1443,15 @@ const RegularSpectrum31f RGBToSpectrumRedIlluminance(RegularSpectrum31f::from_ar
 const RegularSpectrum31f RGBToSpectrumGreenIlluminance(RegularSpectrum31f::from_array(RGBToSpectrumGreenIlluminanceTab));
 const RegularSpectrum31f RGBToSpectrumBlueIlluminance(RegularSpectrum31f::from_array(RGBToSpectrumBlueIlluminanceTab));
 
+// Get color matching functions normalizer
+float get_cmf_normalizer(Color4f cmf[32])
```
Comment (dictoon) [inline src/appleseed/foundation/image/colorspace.cpp:1447]: Make argument `const`.

Comment (dictoon) [reply src/appleseed/foundation/image/colorspace.cpp:1447]: Wrap into anonymous namespace.

File: src/appleseed.studio/utility/inputwidgetproxies.cpp [hunk 29]
```diff
@@ -308,7 +308,7 @@ namespace
                     output_spectrum);
 
                 Color3f ciexyz;
-                spectrum_to_ciexyz_standard(&output_spectrum[0], &ciexyz[0]);
+                spectral_reflectance_to_ciexyz_standard(&output_spectrum[0], &ciexyz[0]);
```
Comment (dictoon) [inline src/appleseed.studio/utility/inputwidgetproxies.cpp:311]: Add a `// todo:` because this isn't correct when picking a color for an emitter.

File: src/appleseed/foundation/meta/benchmarks/benchmark_colorspace.cpp [hunk 30]
```diff
@@ -81,7 +81,7 @@ BENCHMARK_SUITE(Foundation_Image_ColorSpace)
 
     BENCHMARK_CASE_F(SpectrumToCIEXYZ, SpectrumToCIEXYZFixture)
```
Comment (dictoon) [inline src/appleseed/foundation/meta/benchmarks/benchmark_colorspace.cpp:82]: Update case name: `SpectralReflectanceToCIEXYZ`.

File: src/appleseed/foundation/meta/tests/test_colorspace.cpp [hunk 31]
```diff
@@ -482,7 +482,7 @@ TEST_SUITE(Foundation_Image_ColorSpace)
     {
         const RegularSpectrum31f spectrum = get_white_spectrum();
         const LightingConditions lighting_conditions(IlluminantCIED65, XYZCMFCIE19312Deg);
-        const Color3f ciexyz = spectrum_to_ciexyz<float>(lighting_conditions, spectrum);
```
Comment (dictoon) [inline src/appleseed/foundation/meta/tests/test_colorspace.cpp:485]: Update test case name.

File: src/appleseed/renderer/modeling/color/colorspace.cpp [hunk 32]
```diff
@@ -34,7 +34,7 @@ using namespace foundation;
 namespace renderer
 {
 
-LightingConditions g_std_lighting_conditions;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/color/colorspace.cpp:37]: Revert.

File: src/appleseed/renderer/modeling/color/colorspace.h (hunk 14 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/color/colorspace.h:39]: Nice catch!

File: src/appleseed/renderer/modeling/color/colorspace.h [hunk 33]
```diff
@@ -34,10 +34,11 @@
...
 {
 
+
 //
-// Standard lighting conditions (D65 illuminant, CIE 1964 10-deg color matching functions).
+// Standard lighting conditions (D65 illuminant, CIE 1931 2-deg color matching functions).
 //
 
-extern foundation::LightingConditions g_std_lighting_conditions;
+extern foundation::LightingConditions       g_std_lighting_conditions;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/color/colorspace.h:42]: Revert.

File: src/appleseed/renderer/modeling/input/colorsource.cpp [hunk 34]
```diff
@@ -109,14 +116,23 @@ void ColorSource::initialize_from_spectrum(const ColorEntity& color_entity)
...
         s);
 
-    m_linear_rgb =
-        ciexyz_to_linear_rgb(
-            spectrum_to_ciexyz<float>(g_std_lighting_conditions, s));
-
-    m_spectrum.set(s, g_std_lighting_conditions, Spectrum::Reflectance);
+    if (intent == Spectrum::Reflectance)
+    {
+        m_spectrum.set(s, g_std_lighting_conditions, intent);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/input/colorsource.cpp:121]: This line can be moved out of the `if`...

File: src/appleseed/renderer/modeling/input/colorsource.cpp [hunk 35]
```diff
@@ -109,14 +116,23 @@ void ColorSource::initialize_from_spectrum(const ColorEntity& color_entity)
...
+        m_linear_rgb =
+            ciexyz_to_linear_rgb(
+                spectral_reflectance_to_ciexyz<float>(g_std_lighting_conditions, s));
+    }
+    else
+    {
+        m_spectrum.set(s, g_std_lighting_conditions, intent);
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/input/colorsource.cpp:128]: ...since it's the same as this one.

File: src/appleseed/renderer/utility/dynamicspectrum.h [hunk 36]
```diff
@@ -470,25 +489,47 @@ inline const T& DynamicSpectrum<T, N>::operator[](const size_t i) const
...
+                  foundation::spectral_reflectance_to_ciexyz<T>(lighting_conditions, *this));
+}
+
+template <typename T, size_t N>
+inline foundation::Color<T, 3> DynamicSpectrum<T, N>::illuminance_to_rgb(
+    const foundation::LightingConditions& lighting_conditions) const
+{
+    return
+        s_mode == RGB
+        ? foundation::Color<T, 3>(m_samples[0], m_samples[1], m_samples[2])
```
Comment (dictoon) [inline src/appleseed/renderer/utility/dynamicspectrum.h:508]: Fix indentation (check the previous function for how to indent such code).

File: src/appleseed/renderer/utility/dynamicspectrum.h [hunk 37]
```diff
@@ -470,25 +489,47 @@ inline const T& DynamicSpectrum<T, N>::operator[](const size_t i) const
...
+            : foundation::spectral_reflectance_to_ciexyz<T>(lighting_conditions, *this);
+}
+
+template <typename T, size_t N>
+inline foundation::Color<T, 3> DynamicSpectrum<T, N>::illuminance_to_ciexyz(
+    const foundation::LightingConditions& lighting_conditions) const
+{
+    return
+        s_mode == RGB
+        ? linear_rgb_to_ciexyz(
```
Comment (dictoon) [inline src/appleseed/renderer/utility/dynamicspectrum.h:530]: Fix indentation.

File: src/appleseed/foundation/image/colorspace.h (hunk 26 above)
Comment (mororo250) [reply src/appleseed/foundation/image/colorspace.h:165]: Sorry about that. This comment was for the normalizer. I forgot to change it.

**Comment by oktomus** [comment]
This PR is most likely the cause of all the differences we got lately in the test suite: http://testsuite.appleseedhq.net:8080/current-job/sandbox/tests/test%20scenes/report.html
